source/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller.h

```diff
@@ -0,0 +1,150 @@
+#include "envoy/event/dispatcher.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/filters/http/adaptive_concurrency/concurrency_controller/concurrency_controller.h"
+
+#include "absl/base/thread_annotations.h"
+#include "absl/synchronization/mutex.h"
+#include "circllhist.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdaptiveConcurrency {
+namespace ConcurrencyController {
+
+/**
+ * All stats for the gradient controller.
+ */
+// clang-format off
+#define ALL_GRADIENT_CONTROLLER_STATS(COUNTER, GAUGE) \
+  GAUGE(concurrency_limit, Accumulate)  \
+  GAUGE(gradient, Accumulate)  \
+  GAUGE(burst_queue_size, Accumulate)  \
+  GAUGE(rq_outstanding, Accumulate)  \
+  GAUGE(min_rtt_msecs, Accumulate)
+// clang-format on
+
+/**
+ * Wrapper struct for gradient controller stats. @see stats_macros.h
+ */
+struct GradientControllerStats {
+  ALL_GRADIENT_CONTROLLER_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+class GradientControllerConfig {
+public:
+  GradientControllerConfig(
+      const envoy::config::filter::http::adaptive_concurrency::v2alpha::GradientControllerConfig&
+          proto_config);
+
+  // Accessors.
+  std::chrono::milliseconds min_rtt_calc_interval() const { return min_rtt_calc_interval_; }
+  std::chrono::milliseconds sample_rtt_calc_interval() const { return sample_rtt_calc_interval_; }
+  uint64_t max_concurrency_limit() const { return max_concurrency_limit_; }
+  int min_rtt_aggregate_request_count() const { return min_rtt_aggregate_request_count_; }
+  double max_gradient() const { return max_gradient_; }
+  double sample_aggregate_percentile() const { return sample_aggregate_percentile_; }
+
+private:
+  // The measured request round-trip time under ideal conditions.
+  std::chrono::milliseconds min_rtt_calc_interval_;
+
+  // The measured sample round-trip time from the previous time window.
+  std::chrono::milliseconds sample_rtt_calc_interval_;
+
+  // The maximum allowed concurrency value.
+  uint64_t max_concurrency_limit_;
+
+  // The number of requests to aggregate/sample during the minRTT recalculation.
+  int min_rtt_aggregate_request_count_;
+
+  // The maximum value the gradient may take.
+  double max_gradient_;
+
+  // The percentile value considered when processing samples.
+  double sample_aggregate_percentile_;
+};
+using GradientControllerConfigSharedPtr = std::shared_ptr<GradientControllerConfig>;
+
+/**
+ * A concurrency controller that implements the a variation of the Gradient algorithm to control the
+ * allowed concurrency window.
+ *
+ * An ideal round-trip time (minRTT) is measured periodically by only allowing a single outstanding
+ * request at a time and measuring the round-trip time to the listener. This information is then
+ * used in the calculation of a number called the gradient, using time-sampled latencies
+ * (sampleRTT):
+ *
+ *     gradient = minRTT / sampleRTT
```

Comment:
Can you describe a bit more about how sample RTT is computed? I assume that is done after computing minRTT? Basically it's done between the minRTT calculations? Or does it include the minRTT calculations? Etc.